types/trace.go

```diff
@@ -0,0 +1,132 @@
+package types
+
+import (
+	"encoding/json"
+	"errors"
+	"time"
+)
+
+// TraceSession model.
+type TraceSession struct {
+	ID         string    `json:"id" yaml:"id"`
+	PipelineID string    `json:"pipelineID" yaml:"pipelineID"`
+	Plugins    []string  `json:"plugins" yaml:"plugins"`
+	Lifespan   Duration  `json:"lifespan" yaml:"lifespan"`
+	CreatedAt  time.Time `json:"createdAt" yaml:"createdAt"`
+	UpdatedAt  time.Time `json:"updatedAt" yaml:"updatedAt"`
+}
+
+// Active tells whether a session is still within its lifespan.
+func (ts TraceSession) Active() bool {
+	return ts.CreatedAt.Add(time.Duration(ts.Lifespan)).After(time.Now())
+}
+
+// CreateTraceSession request payload for creating a new trace session.
+type CreateTraceSession struct {
+	Plugins  []string `json:"plugins"`
+	Lifespan Duration `json:"lifespan"`
+}
+
+// TraceSessionsParams request payload for querying trace sessions.
+type TraceSessionsParams struct {
+	Last   *uint64
+	Before *string
+}
+
+// TraceSessions paginated list.
+type TraceSessions struct {
+	Items     []TraceSession `json:"items"`
+	EndCursor *string        `json:"endCursor"`
+}
+
+// UpdateTraceSession request payload for updating a trace session.
+type UpdateTraceSession struct {
+	Plugins  *[]string `json:"plugins"`
+	Lifespan *Duration `json:"lifespan"`
+}
+
+// TraceRecord model.
+type TraceRecord struct {
+	ID        string    `json:"id" yaml:"id"`
+	SessionID string    `json:"sessionID" yaml:"sessionID"`
+	CreatedAt time.Time `json:"createdAt" yaml:"createdAt"`
+
+	// fluent-bit data from here on.
+	Kind           TraceRecordKind `json:"type" yaml:"type"`
+	TraceID        string          `json:"trace_id" yaml:"trace_id"`
+	StartTime      time.Time       `json:"start_time" yaml:"start_time"`
+	EndTime        time.Time       `json:"end_time" yaml:"end_time"`
+	PluginInstance string          `json:"plugin_instance" yaml:"plugin_instance"`
+	PluginAlias    string          `json:"plugin_alias" yaml:"plugin_alias"`
+	ReturnCode     int             `json:"return_code" yaml:"return_code"`
+	// Records array, each record is a JSON object,
+	// warranted to have a flb_time `timestamp` field.
+	Records json.RawMessage `json:"records" yaml:"records"`
```

Comment:
The records have the following format, in case you want to implement it:

```json
{
    "records": {
        "timestamp"=>1658952151,
        "record": {
            "message": "dummy"
        }
    }
}
```

The actual record field in each record is a variant though, usually a key/value map. It might not be worth it if we are just handing the records off directly to core-ui to interpret it though.

Reply:
Wait. I thought it was an array. Like this:
```
{
	"id": "some_record_id",
	"sessionID": "some_session_id",
	"type": 1,
	"trace_id": "some_trace_id",
	"start_time": "2009-11-10T23:00:00Z",
	"end_time": "2009-11-10T23:00:00Z",
	"plugin_instance": "some_plugin_instance",
	"plugin_alias": "some_plugin_alias",
	"return_code": 33,
	"records": [
		{
			"timestamp": "2009-11-10T23:00:00Z",
			"some_other_value": "some_record_value"
		}
	],
	"createdAt": "2009-11-10T23:00:00Z"
}
```

Reply:
So the `timestamp` value is numeric too. Unix timestamp I guess.

Reply:
@pwhelan I need to get the correct shape, cause we do a small validation here https://github.com/calyptia/cloud/blob/46894872fe599795a8bd79fcf11312da55ec31b1/trace.go#L255-L262

Reply:
The format I specified before is the correct one and was taken from this guide: https://github.com/calyptia/guidelines/blob/main/tap/demo-usage.md. The structure of `record` will vary depending on the actual input and its content.

@nicolasparada said:
> So the timestamp value is numeric too. Unix timestamp I guess thinking

It is for now. In the future it may change to be a Unix timestamp but with subseconds represented via decimals.

Reply:
@pwhelan so only to be sure.

```json
{
	"id": "some_record_id",
	"sessionID": "some_session_id",
	"type": 1,
	"trace_id": "some_trace_id",
	"start_time": "2009-11-10T23:00:00Z",
	"end_time": "2009-11-10T23:00:00Z",
	"plugin_instance": "some_plugin_instance",
	"plugin_alias": "some_plugin_alias",
	"return_code": 33,
	"records": {
		"timestamp": 125436346345,
		"record": {
			"key": "val",
			"count": 1,
			"attr": {
				"one": "two"
			}
		}
	},
	"createdAt": "2009-11-10T23:00:00Z"
}
```

`records.record` being a free-form JSON object.

Reply:
> records.record being a free-form JSON object.

Exactly. 👍